README
NAME
    URI::Fetch - Smart URI fetching/caching

SYNOPSIS
        use URI::Fetch;

        ## Simple fetch.
        my $res = URI::Fetch->fetch('http://example.com/atom.xml')
            or die URI::Fetch->errstr;

        ## Fetch using specified ETag and Last-Modified headers.
        $res = URI::Fetch->fetch('http://example.com/atom.xml',
                ETag => '123-ABC',
                LastModified => time - 3600,
        )
            or die URI::Fetch->errstr;

        ## Fetch using an on-disk cache that URI::Fetch manages for you.
        my $cache = Cache::File->new( cache_root => '/tmp/cache' );
        $res = URI::Fetch->fetch('http://example.com/atom.xml',
                Cache => $cache
        )
            or die URI::Fetch->errstr;

DESCRIPTION
    *URI::Fetch* is a smart client for fetching HTTP pages, notably
    syndication feeds (RSS, Atom, and others), in an intelligent, bandwidth-
    and time-saving way. That means:

    *   GZIP support

        If you have *Compress::Zlib* installed, *URI::Fetch* will
        automatically try to download a compressed version of the content,
        saving bandwidth (and time).

    *   *Last-Modified* and *ETag* support

        If you use a local cache (see the *Cache* parameter to *fetch*),
        *URI::Fetch* will keep track of the *Last-Modified* and *ETag*
        headers from the server, allowing you to only download pages that
        have been modified since the last time you checked.

    *   Proper understanding of HTTP error codes

        Certain HTTP error codes are special, particularly when fetching
        syndication feeds, and well-written clients should pay special
        attention to them. *URI::Fetch* can only do so much for you in this
        regard, but it gives you the tools to be a well-written client.

        The response from *fetch* gives you the raw HTTP response code,
        along with special handling of 4 codes:

        *   200 (OK)

            Signals that the content of a page/feed was retrieved
            successfully.

        *   301 (Moved Permanently)

            Signals that a page/feed has moved permanently, and that your
            database of feeds should be updated to reflect the new URI.

        *   304 (Not Modified)

            Signals that a page/feed has not changed since it was last
            fetched.

        *   410 (Gone)

            Signals that a page/feed is gone and will never be coming back,
            so you should stop trying to fetch it.

USAGE
  URI::Fetch->fetch($uri, %param)
    Fetches a page identified by the URI *$uri*.

    On success, returns a *URI::Fetch::Response* object; on failure, returns
    "undef".

    *%param* can contain:

    *   LastModified

    *   ETag

        *LastModified* and *ETag* can be supplied to force the server to
        only return the full page if it's changed since the last request. If
        you're writing your own feed client, this is recommended practice,
        because it limits both your bandwidth use and the server's.

        If you'd rather not have to store the *LastModified* time and *ETag*
        yourself, see the *Cache* parameter below (and the SYNOPSIS above).

    *   Cache

        If you'd like *URI::Fetch* to cache responses between requests,
        provide the *Cache* parameter with an object supporting the Cache
        API (e.g. *Cache::File*, *Cache::Memory*). Specifically, an object
        that supports "$cache->get($key)" and "$cache->set($key, $value,
        $expires)".

        If supplied, *URI::Fetch* will store the page content, ETag, and
        last-modified time of the response in the cache, and will pull the
        content from the cache on subsequent requests if the page returns a
        Not-Modified response.

    *   UserAgent

        Optional. You may provide your own LWP::UserAgent instance. Look
        into LWPx::ParanoidUserAgent if you're fetching URLs given to you by
        possibly malicious parties.

    *   NoNetwork

        Optional. Controls the interaction between the cache and HTTP
        requests with If-Modified-Since/If-None-Match headers. Possible
        behaviors are:

        false (default)
            If a page is in the cache, the origin HTTP server is always
            checked for a fresher copy with an If-Modified-Since and/or
            If-None-Match header.

        1   If set to 1, the origin HTTP is never contacted, regardless of
            the page being in cache or not. If the page is missing from
            cache, the fetch method will return undef. If the page is in
            cache, that page will be returned, no matter how old it is. Note
            that setting this option means the URI::Fetch::Response object
            will never have the http_response member set.

        "N", where N > 1
            The origin HTTP server is not contacted if the page is in cache
            and the cached page was inserted in the last N seconds. If the
            cached copy is older than N seconds, a normal HTTP request (full
            or cache check) is done.

    *   ContentAlterHook

        Optional. A subref that gets called with a scalar reference to your
        content so you can modify the content before it's returned and
        before it's put in cache.

        For instance, you may want to only cache the <head> section of an
        HTML document, or you may want to take a feed URL and cache only a
        pre-parsed version of it. If you modify the scalarref given to your
        hook and change it into a hashref, scalarref, or some blessed
        object, that same value will be returned to you later on
        not-modified responses.

    *   CacheEntryGrep

        Optional. A subref that gets called with the *URI::Fetch::Response*
        object about to be cached (with the contents already possibly
        transformed by your "ContentAlterHook"). If your subref returns
        true, the page goes into the cache. If false, it doesn't.

    *   Freeze

    *   Thaw

        Optional. Subrefs that get called to serialize and deserialize,
        respectively, the data that will be cached. The cached data should
        be assumed to be an arbitrary Perl data structure, containing
        (potentially) references to arrays, hashes, etc.

        Freeze should serialize the structure into a scalar; Thaw should
        deserialize the scalar into a data structure.

        By default, *Storable* will be used for freezing and thawing the
        cached data structure.

    *   ForceResponse

        Optional. A boolean that indicates a *URI::Fetch::Response* should
        be returned regardless of the HTTP status. By default "undef" is
        returned when a response is not a "success" (200 codes) or one of
        the recognized HTTP status codes listed above. The HTTP status
        message can then be retreived using the "errstr" method on the
        class.

LICENSE
    *URI::Fetch* is free software; you may redistribute it and/or modify it
    under the same terms as Perl itself.

AUTHOR & COPYRIGHT
    Except where otherwise noted, *URI::Fetch* is Copyright 2004 Benjamin
    Trott, ben+cpan@stupidfool.org. All rights reserved.



Changes
Revision history for Perl module URI::Fetch

0.14 2021-05-26 NEILB
    - When Benjamin created this, he had a server set up for doing the
      tests, but that's gone now. So to stop all the fails, I'm skipping
      the live tests for now. Could do with mocking, but this is a stop-gap.

0.13_01 2021-05-25 NEILB
    - Changed the test suite to use some online HTTP status code test servers.
      Let's see how reliable they are.

0.13 2016-07-02 NEILB
    - Switched to using Test::RequiresInternet rather than the hand-rolled
      online() function that was duplicated in 2 tests. Plus it was testing
      whether google.com was contactable, and not the site used in the tests.

0.12 2016-06-23 NEILB
    - Travis support added by Jason Hall.
    - Added list of contributors to the doc.

0.11 2015-07-04 NEILB
    - Added [MetaJSON] to dist.ini, so release will include META.json.
      RT#105632 from ETHER++

0.10 2014-06-23
    - CPAN Testers looking good after previous developer release.
    - Added github repo to pod

0.09_01 2014-06-13
    - If you've got caching enabled, and get a 304 response (Not Modified)
      with content (from the cache), then is_success() returns true.
      Suggested in RT#75665
    - Caching now done under the original url rather than the sanitised
      version of it. Bug report and patch from Mario Domgoergen RT#39820
    - Switched to Dist::Zilla
    - Reformatted Changes as per CPAN::Changes::Spec

0.09  2011-01-28
    - Use $ua->env_proxy to load local proxy settings. (RT 53819)
    - Skip tests if we don't have a network connection. (RT 28388, 59694)
    - Removed sign() and auto_install() from Makefile.PL.
    - Removed magic svn keywords.
    - Added author tests (xt/) and modified SYNOPSIS for all modules to
      make them pass the compilation test.

0.08  2006-07-24
    - Don't overwrite the User-Agent field if the caller has set it and
      provided the UserAgent argument. Thanks to Tatsuhiko Miyagawa
      for the patch.

0.071 2006-06-25
    - Fixed broken is_error (broken in 0.07). Thanks to Tatsuhiko for the
      patch.

0.07  2006-06-18
    - Use $res->header('Content-Type') instead of $res->content_type, since
      the latter can return an array. Thanks to Tatsuhiko Miyagawa for the
      patch.
    - $res->is_success, is_error, and is_redirect previously threw an exception
      when called with NoNetwork. They just work as 200 succesful request now,
      when the response is taken back from the cache. Thanks to Tatsuhiko
      Miyagawa for the patch.

0.06  2006-04-09
    - Fixed issue where content-type was not stored in the cache, and was
      thus blank on subsequent requests. Thanks to Tatsuhiko Miyagawa for
      the patch.
    - Fixed issue with caching redirected (304) URIs. Thanks to Tatsuhiko
      Miyagawa for the patch.
    - Added is_error, is_redirect, is_success convenience methods to
      URI::Fetch::Response. Thanks to Tatsuhiko Miyagawa for the patch.

0.05  2006-02-24
    - Added a ForceResponse option, which forces URI::Fetch->fetch to return
      a URI::Fetch::Response object even if the HTTP request fails for an
      unknown reason. Thanks to Tim Appnel for the patch.

0.04  2005-10-09
    - Added Thaw and Freeze options, which allow you to define the
      serialization and deserialization options that are used. Thanks to
      Tim Appnel for the patch.

0.03  2005-05-27
    - Added a NoNetwork option, allowing fetch to trust the cache completely
      and skip the HTTP request; or the option to do this only if the version
      in the cache is older than N seconds. [bradfitz]
    - Added a CacheEntryGrep option, to allow for not caching certain
      responses. [bradfitz]
    - Documentation fixes & clarifications. [bradfitz]

0.02  2005-05-25
    - Let the caller pass in their own UserAgent and ContentAlterHook.
      [bradfitz]
    - Be more strict about invalid parameters. [bradfitz]
    - Documentation fixes for URI::Fetch->fetch in the SYNOPSIS. Thanks
      to Naoya Ito for the note.

0.01  2004-12-31
    - Initial distribution.



inc/Module/Install.pm
#line 1 "/Users/btrott/devel/cpan/uri-fetch/inc/Module/Install.pm - /Library/Perl/5.8.6/Module/Install.pm"
package Module::Install;

use 5.004;
use strict 'vars';

use vars qw{$VERSION};
BEGIN {
    # All Module::Install core packages now require synchronised versions.
    # This will be used to ensure we don't accidentally load old or
    # different versions of modules.
    # This is not enforced yet, but will be some time in the next few
    # releases once we can make sure it won't clash with custom
    # Module::Install extensions.
    $VERSION = '0.57';
}

# inc::Module::Install must be loaded first
unless ( $INC{join('/', inc => split(/::/, __PACKAGE__)).'.pm'} ) {
    die <<"END_DIE";
Please invoke ${\__PACKAGE__} with:

    use inc::${\__PACKAGE__};

not:

    use ${\__PACKAGE__};

END_DIE
}

use Cwd        ();
use File::Find ();
use File::Path ();
use FindBin;

*inc::Module::Install::VERSION = *VERSION;
@inc::Module::Install::ISA     = 'Module::Install';

sub autoload {
    my $self   = shift;
    my $caller = $self->_caller;
    my $cwd    = Cwd::cwd();
    my $sym    = "$caller\::AUTOLOAD";

    $sym->{$cwd} = sub {
        my $pwd = Cwd::cwd();
        if ( my $code = $sym->{$pwd} ) {
            # delegate back to parent dirs
            goto &$code unless $cwd eq $pwd;
        }
        $$sym =~ /([^:]+)$/ or die "Cannot autoload $caller - $sym";
        unshift @_, ($self, $1);
        goto &{$self->can('call')} unless uc($1) eq $1;
    };
}

sub import {
    my $class = shift;
    my $self  = $class->new(@_);

    unless ( -f $self->{file} ) {
        require "$self->{path}/$self->{dispatch}.pm";
        File::Path::mkpath("$self->{prefix}/$self->{author}");
        $self->{admin} = 
          "$self->{name}::$self->{dispatch}"->new(_top => $self);
        $self->{admin}->init;
        @_ = ($class, _self => $self);
        goto &{"$self->{name}::import"};
    }

    *{$self->_caller . "::AUTOLOAD"} = $self->autoload;
    $self->preload;

    # Unregister loader and worker packages so subdirs can use them again
    delete $INC{"$self->{file}"};
    delete $INC{"$self->{path}.pm"};
}

sub preload {
    my ($self) = @_;

    unless ( $self->{extensions} ) {
        $self->load_extensions(
            "$self->{prefix}/$self->{path}", $self
        );
    }

    my @exts = @{$self->{extensions}};
    unless ( @exts ) {
        my $admin = $self->{admin};
        @exts = $admin->load_all_extensions;
    }

    my %seen;
    foreach my $obj ( @exts ) {
        while (my ($method, $glob) = each %{ref($obj) . '::'}) {
            next unless defined *{$glob}{CODE};
            next if $method =~ /^_/;
            next if $method eq uc($method);
            $seen{$method}++;
        }
    }

    my $caller = $self->_caller;
    foreach my $name ( sort keys %seen ) {
        *{"${caller}::$name"} = sub {
            ${"${caller}::AUTOLOAD"} = "${caller}::$name";
            goto &{"${caller}::AUTOLOAD"};
        };
    }
}

sub new {
    my ($class, %args) = @_;

    # ignore the prefix on extension modules built from top level.
    my $base_path = Cwd::abs_path($FindBin::Bin);
    unless ( Cwd::abs_path(Cwd::cwd()) eq $base_path ) {
        delete $args{prefix};
    }

    return $args{_self} if $args{_self};

    $args{dispatch} ||= 'Admin';
    $args{prefix}   ||= 'inc';
    $args{author}   ||= '.author';
    $args{bundle}   ||= 'inc/BUNDLES';
    $args{base}     ||= $base_path;

    $class =~ s/^\Q$args{prefix}\E:://;
    $args{name}     ||= $class;
    $args{version}  ||= $class->VERSION;

    unless ( $args{path} ) {
        $args{path}  = $args{name};
        $args{path}  =~ s!::!/!g;
    }
    $args{file}     ||= "$args{base}/$args{prefix}/$args{path}.pm";

    bless(\%args, $class);
}

sub call {
    my $self   = shift;
    my $method = shift;
    my $obj    = $self->load($method) or return;

    unshift @_, $obj;
    goto &{$obj->can($method)};
}

sub load {
    my ($self, $method) = @_;

    $self->load_extensions(
        "$self->{prefix}/$self->{path}", $self
    ) unless $self->{extensions};

    foreach my $obj (@{$self->{extensions}}) {
        return $obj if $obj->can($method);
    }

    my $admin = $self->{admin} or die <<"END_DIE";
The '$method' method does not exist in the '$self->{prefix}' path!
Please remove the '$self->{prefix}' directory and run $0 again to load it.
END_DIE

    my $obj = $admin->load($method, 1);
    push @{$self->{extensions}}, $obj;

    $obj;
}

sub load_extensions {
    my ($self, $path, $top_obj) = @_;

    unshift @INC, $self->{prefix}
        unless grep { $_ eq $self->{prefix} } @INC;

    local @INC = ($path, @INC);
    foreach my $rv ($self->find_extensions($path)) {
        my ($file, $pkg) = @{$rv};
        next if $self->{pathnames}{$pkg};

        local $@;
        my $new = eval { require $file; $pkg->can('new') };
        unless ( $new ) {
            warn $@ if $@;
            next;
        }
        $self->{pathnames}{$pkg} = delete $INC{$file};
        push @{$self->{extensions}}, &{$new}($pkg, _top => $top_obj );
    }

    $self->{extensions} ||= [];
}

sub find_extensions {
    my ($self, $path) = @_;

    my @found;
    File::Find::find( sub {
        my $file = $File::Find::name;
        return unless $file =~ m!^\Q$path\E/(.+)\.pm\Z!is;
        return if $1 eq $self->{dispatch};

        $file = "$self->{path}/$1.pm";
        my $pkg = "$self->{name}::$1"; $pkg =~ s!/!::!g;
        push @found, [ $file, $pkg ];
    }, $path ) if -d $path;

    @found;
}

sub _caller {
    my $depth  = 0;
    my $caller = caller($depth);

    while ($caller eq __PACKAGE__) {
        $depth++;
        $caller = caller($depth);
    }

    $caller;
}

1;


lib/URI/Fetch/Response.pm
package URI::Fetch::Response;

use strict;
use warnings;

sub new {
    my $class = shift;
    my $feed = bless { }, $class;
    $feed;
}

sub _var {
    my $feed = shift;
    my $var = shift;
    $feed->{$var} = shift if @_;
    $feed->{$var};
}

sub status        { shift->_var('status',        @_) }
sub http_status   { shift->_var('http_status',   @_) }
sub http_response { shift->_var('http_response', @_) }
sub etag          { shift->_var('etag',          @_) }
sub last_modified { shift->_var('last_modified', @_) }
sub uri           { shift->_var('uri',           @_) }
sub content       { shift->_var('content',       @_) }
sub content_type  { shift->_var('content_type',  @_) }

sub is_success  {
    my $response = shift;
    if ($response->http_response) {
        return 1 if $response->http_response->code == 304
                 && defined($response->content);
        return $response->http_response->is_success;
    }
    return 1;
}

sub is_redirect {
    my $response = shift;
    return $response->http_response->is_redirect if $response->http_response;
    return;
}

sub is_error    {
    my $response = shift;
    return $response->http_response->is_error if $response->http_response;
    return;
}

1;
__END__

=head1 NAME

URI::Fetch::Response - Feed response for URI::Fetch

=head1 SYNOPSIS

    use URI::Fetch;
    my $res = URI::Fetch->fetch('http://example.com/atom.xml')
        or die URI::Fetch->errstr;
    print $res->content;

=head1 DESCRIPTION

I<URI::Fetch::Response> encapsulates the response from fetching a feed
using I<URI::Fetch>.

=head1 USAGE

=head2 $res->content

The contents of the feed.

=head2 $res->uri

The URI of the feed. If the feed was moved, this reflects the new URI;
otherwise, it will match the URI that you passed to I<fetch>.

=head2 $res->etag

The ETag that was returned in the response, if any.

=head2 $res->last_modified

The Last-Modified date (in seconds since the epoch) that was returned in
the response, if any.

=head2 $res->status

The status of the response, which will match one of the following
enumerations:

=over 4

=item * URI::Fetch::URI_OK()

=item * URI::Fetch::URI_MOVED_PERMANENTLY()

=item * URI::Fetch::URI_GONE()

=item * URI::Fetch::URI_NOT_MODIFIED()

=back

=head2 $res->http_status

The HTTP status code from the response.

=head2 $res->http_response

The I<HTTP::Response> object returned from the fetch.

=head2 $res->is_success

=head2 $res->is_redirect

=head2 $res->is_error

Wrappers around the C<$res-E<gt>response> methods of the same name, for
convenience.

B<Note:> there is one difference from the behaviour of L<HTTP::Response>.
If you are using a cache and get a 304 response, but the data is retrieved
from the cache, then C<is_success> will return true,
because C<res-E<gt>content> is usable.

=head2 $res->content_type

The Content-Type header from the response.

=head1 AUTHOR & COPYRIGHT

Please see the I<URI::Fetch> manpage for author, copyright, and license
information.

=cut


lib/URI/Fetch.pm
package URI::Fetch;

use 5.008001;
use strict;
use warnings;

use base qw( Class::ErrorHandler );

use LWP::UserAgent;
use Carp qw( croak );
use URI;
use URI::Fetch::Response;

our $HAS_ZLIB;
BEGIN {
    $HAS_ZLIB = eval "use Compress::Zlib (); 1;";
}

use constant URI_OK                => 200;
use constant URI_MOVED_PERMANENTLY => 301;
use constant URI_NOT_MODIFIED      => 304;
use constant URI_GONE              => 410;

sub fetch {
    my $class = shift;
    my($uri, %param) = @_;

    # get user parameters
    my $cache        = delete $param{Cache};
    my $ua           = delete $param{UserAgent};
    my $p_etag       = delete $param{ETag};
    my $p_lastmod    = delete $param{LastModified};
    my $content_hook = delete $param{ContentAlterHook};
    my $p_no_net     = delete $param{NoNetwork};
    my $p_cache_grep = delete $param{CacheEntryGrep};
    my $freeze       = delete $param{Freeze};
    my $thaw         = delete $param{Thaw};
    my $force        = delete $param{ForceResponse};
    croak("Unknown parameters: " . join(", ", keys %param))
        if %param;

    my $ref;
    if ($cache) {
        unless ($freeze && $thaw) {
            require Storable;
            $thaw = \&Storable::thaw;
            $freeze = \&Storable::freeze;
        }
        if (my $blob = $cache->get($uri)) {
            $ref = $thaw->($blob);
        }
    }

    # NoNetwork support (see pod docs below for logic clarification)
    if ($p_no_net) {
        croak("Invalid NoNetworkValue (negative)") if $p_no_net < 0;
        if ($ref && ($p_no_net == 1 || $ref->{CacheTime} > time() - $p_no_net)) {

            my $fetch = URI::Fetch::Response->new;
            $fetch->status(URI_OK);
            $fetch->content($ref->{Content});
            $fetch->etag($ref->{ETag});
            $fetch->last_modified($ref->{LastModified});
            $fetch->content_type($ref->{ContentType});
            return $fetch;
        }
        return undef if $p_no_net == 1;
    }

    $ua ||= do {
        my $ua = LWP::UserAgent->new;
        $ua->agent(join '/', $class, $class->VERSION);
        $ua->env_proxy;
        $ua;
    };

    my $req = HTTP::Request->new(GET => $uri);
    if ($HAS_ZLIB) {
        $req->header('Accept-Encoding', 'gzip');
    }
    if (my $etag = ($p_etag || $ref->{ETag})) {
        $req->header('If-None-Match', $etag);
    }
    if (my $ts = ($p_lastmod || $ref->{LastModified})) {
        $req->if_modified_since($ts);
    }

    my $res = $ua->request($req);
    my $fetch = URI::Fetch::Response->new;
    $fetch->uri($uri);
    $fetch->http_status($res->code);
    $fetch->http_response($res);
    $fetch->content_type($res->header('Content-Type'));
    if ($res->previous && $res->previous->code == HTTP::Status::RC_MOVED_PERMANENTLY()) {
        $fetch->status(URI_MOVED_PERMANENTLY);
        $fetch->uri($res->previous->header('Location'));
    } elsif ($res->code == HTTP::Status::RC_GONE()) {
        $fetch->status(URI_GONE);
        $fetch->uri(undef);
        return $fetch;
    } elsif ($res->code == HTTP::Status::RC_NOT_MODIFIED()) {
        $fetch->status(URI_NOT_MODIFIED);
        $fetch->content($ref->{Content});
        $fetch->etag($ref->{ETag});
        $fetch->last_modified($ref->{LastModified});
        $fetch->content_type($ref->{ContentType});
        return $fetch;
    } elsif (!$res->is_success) {
        return $force ? $fetch : $class->error($res->message);
        
    } else {
        $fetch->status(URI_OK);
    }
    $fetch->last_modified($res->last_modified);
    $fetch->etag($res->header('ETag'));
    my $content = $res->content;
    if ($res->content_encoding && $res->content_encoding eq 'gzip') {
        $content = Compress::Zlib::memGunzip($content);
    }

    # let caller-defined transform hook modify the result that'll be
    # cached.  perhaps the caller only wants the <head> section of
    # HTML, or wants to change the content to a parsed datastructure
    # already serialized with Storable.
    if ($content_hook) {
        croak("ContentAlterHook is not a subref") unless ref $content_hook eq "CODE";
        $content_hook->(\$content);
    }

    $fetch->content($content);

    # cache by default, if there's a cache.  but let callers cancel
    # the cache action by defining a cache grep hook
    if ($cache &&
        ($p_cache_grep ? $p_cache_grep->($fetch) : 1)) {

        $cache->set($uri, $freeze->({
            ETag         => $fetch->etag,
            LastModified => $fetch->last_modified,
            Content      => $fetch->content,
            CacheTime    => time(),
            ContentType  => $fetch->content_type,
        }));
    }
    $fetch;
}

1;
__END__

=head1 NAME

URI::Fetch - Smart URI fetching/caching

=head1 SYNOPSIS

    use URI::Fetch;

    ## Simple fetch.
    my $res = URI::Fetch->fetch('http://example.com/atom.xml')
        or die URI::Fetch->errstr;
    do_something($res->content) if $res->is_success;

    ## Fetch using specified ETag and Last-Modified headers.
    $res = URI::Fetch->fetch('http://example.com/atom.xml',
            ETag => '123-ABC',
            LastModified => time - 3600,
    )
        or die URI::Fetch->errstr;

    ## Fetch using an on-disk cache that URI::Fetch manages for you.
    my $cache = Cache::File->new( cache_root => '/tmp/cache' );
    $res = URI::Fetch->fetch('http://example.com/atom.xml',
            Cache => $cache
    )
        or die URI::Fetch->errstr;

=head1 DESCRIPTION

I<URI::Fetch> is a smart client for fetching HTTP pages, notably
syndication feeds (RSS, Atom, and others), in an intelligent,
bandwidth- and time-saving way. That means:

=over 4

=item * GZIP support

If you have I<Compress::Zlib> installed, I<URI::Fetch> will automatically
try to download a compressed version of the content, saving bandwidth (and
time).

=item * I<Last-Modified> and I<ETag> support

If you use a local cache (see the I<Cache> parameter to I<fetch>),
I<URI::Fetch> will keep track of the I<Last-Modified> and I<ETag> headers
from the server, allowing you to only download pages that have been
modified since the last time you checked.

=item * Proper understanding of HTTP error codes

Certain HTTP error codes are special, particularly when fetching syndication
feeds, and well-written clients should pay special attention to them.
I<URI::Fetch> can only do so much for you in this regard, but it gives
you the tools to be a well-written client.

The response from I<fetch> gives you the raw HTTP response code, along with
special handling of 4 codes:

=over 4

=item * 200 (OK)

Signals that the content of a page/feed was retrieved
successfully.

=item * 301 (Moved Permanently)

Signals that a page/feed has moved permanently, and that
your database of feeds should be updated to reflect the new
URI.

=item * 304 (Not Modified)

Signals that a page/feed has not changed since it was last
fetched.

=item * 410 (Gone)

Signals that a page/feed is gone and will never be coming back,
so you should stop trying to fetch it.

=back

=back

=head2 Change from 0.09

If you make a request using a cache and get back a 304 response code
(Not Modified), then if the content was returned from the cache,
then C<is_success()> will return true, and C<$response-E<gt>content>
will contain the cached content.

I think this is the right behaviour, given the philosophy of C<URI::Fetch>,
but please let me (NEILB) know if you disagree.

=head1 USAGE

=head2 URI::Fetch->fetch($uri, %param)

Fetches a page identified by the URI I<$uri>.

On success, returns a I<URI::Fetch::Response> object; on failure, returns
C<undef>.

I<%param> can contain:

=over 4

=item * LastModified

=item * ETag

I<LastModified> and I<ETag> can be supplied to force the server to only
return the full page if it's changed since the last request. If you're
writing your own feed client, this is recommended practice, because it
limits both your bandwidth use and the server's.

If you'd rather not have to store the I<LastModified> time and I<ETag>
yourself, see the I<Cache> parameter below (and the L<SYNOPSIS> above).

=item * Cache

If you'd like I<URI::Fetch> to cache responses between requests, provide
the I<Cache> parameter with an object supporting the L<Cache> API (e.g.
I<Cache::File>, I<Cache::Memory>). Specifically, an object that supports
C<$cache-E<gt>get($key)> and C<$cache-E<gt>set($key, $value, $expires)>.

If supplied, I<URI::Fetch> will store the page content, ETag, and
last-modified time of the response in the cache, and will pull the
content from the cache on subsequent requests if the page returns a
Not-Modified response.

=item * UserAgent

Optional.  You may provide your own LWP::UserAgent instance.  Look
into L<LWPx::ParanoidUserAgent> if you're fetching URLs given to you
by possibly malicious parties.

=item * NoNetwork

Optional.  Controls the interaction between the cache and HTTP
requests with If-Modified-Since/If-None-Match headers.  Possible
behaviors are:

=over

=item false (default)

If a page is in the cache, the origin HTTP server is always checked
for a fresher copy with an If-Modified-Since and/or If-None-Match
header.

=item C<1>

If set to C<1>, the origin HTTP is never contacted, regardless of the
page being in cache or not.  If the page is missing from cache, the
fetch method will return undef.  If the page is in cache, that page
will be returned, no matter how old it is.  Note that setting this
option means the L<URI::Fetch::Response> object will never have the
http_response member set.

=item C<N>, where N E<gt> 1

The origin HTTP server is not contacted B<if> the page is in cache
B<and> the cached page was inserted in the last N seconds.  If the
cached copy is older than N seconds, a normal HTTP request (full or
cache check) is done.

=back

=item * ContentAlterHook

Optional.  A subref that gets called with a scalar reference to your
content so you can modify the content before it's returned and before
it's put in cache.

For instance, you may want to only cache the E<lt>headE<gt> section of
an HTML document, or you may want to take a feed URL and cache only a
pre-parsed version of it.  If you modify the scalarref given to your
hook and change it into a hashref, scalarref, or some blessed object,
that same value will be returned to you later on not-modified
responses.

=item * CacheEntryGrep

Optional.  A subref that gets called with the I<URI::Fetch::Response>
object about to be cached (with the contents already possibly transformed by
your C<ContentAlterHook>).  If your subref returns true, the page goes
into the cache.  If false, it doesn't.

=item * Freeze

=item * Thaw

Optional. Subrefs that get called to serialize and deserialize, respectively,
the data that will be cached. The cached data should be assumed to be an
arbitrary Perl data structure, containing (potentially) references to
arrays, hashes, etc.

Freeze should serialize the structure into a scalar; Thaw should
deserialize the scalar into a data structure.

By default, I<Storable> will be used for freezing and thawing the cached
data structure.

=item * ForceResponse

Optional. A boolean that indicates a I<URI::Fetch::Response>
should be returned regardless of the HTTP status. By
default C<undef> is returned when a response is not a
"success" (200 codes) or one of the recognized HTTP status
codes listed above. The HTTP status message can then be retreived 
using the C<errstr> method on the class.

=back

=head1 REPOSITORY

L<https://github.com/neilbowers/URI-Fetch>

=head1 LICENSE

I<URI::Fetch> is free software; you may redistribute it and/or modify it
under the same terms as Perl itself.

=head1 AUTHOR & COPYRIGHT

Except where otherwise noted, I<URI::Fetch> is Copyright 2004 Benjamin
Trott, ben+cpan@stupidfool.org. All rights reserved.

Currently maintained by Neil Bowers.

=head1 CONTRIBUTORS

=over 4

=item * Tim Appnel

=item * Mario Domgoergen

=item * Karen Etheridge

=item * Brad Fitzpatrick

=item * Jason Hall

=item * Naoya Ito

=item * Tatsuhiko Miyagawa

=back

=cut

